File: src/routes/recognize.ts


import { Router, Request, Response } from 'express';
import multer from 'multer';
import sharp from 'sharp';
import { config } from '../config';
import { RecognizeOptions } from '../types';
import { recognize } from '../services/recognitionPipeline';
import {
  SnapAccountError,
  ImageTooLargeError,
  ImageTooSmallError,
  ImageFormatUnsupportedError,
} from '../errors';

const router = Router();

const upload = multer({
  storage: multer.memoryStorage(),
  limits: {
    fileSize: config.maxImageSize,
  },
  fileFilter: (_req, file, cb) => {
    if ((config.supportedImageTypes as readonly string[]).includes(file.mimetype)) {
      cb(null, true);
    } else {
      cb(new ImageFormatUnsupportedError(file.mimetype) as any);
    }
  },
});

router.post('/recognize', upload.single('image'), async (req: Request, res: Response) => {
  try {
    
    if (!req.file) {
      res.status(400).json({
        status: 'error',
        error: {
          code: 'INVALID_REQUEST',
          message: 'No image provided. Send an image file as multipart form-data with field name "image".',
        },
      });
      return;
    }

    const metadata = await sharp(req.file.buffer).metadata();
    if (
      (metadata.width && metadata.width < config.minImageWidth) ||
      (metadata.height && metadata.height < config.minImageHeight)
    ) {
      throw new ImageTooSmallError(metadata.width || 0, metadata.height || 0);
    }

    let options: RecognizeOptions = {};
    if (req.body?.options) {
      try {
        options = typeof req.body.options === 'string'
          ? JSON.parse(req.body.options)
          : req.body.options;
      } catch {
        
      }
    }

    const result = await recognize(req.file.buffer, {
      ...options,
      returnAlternatives: true,
      maxAlternatives: options.maxAlternatives ?? 20,
    });

    res.json({
      status: 'success',
      data: {
        account_number: result.accountNumber,
        bank_code: result.bank?.code || null,
        bank_name: result.bank?.name || null,
        account_name: result.accountHolder?.name || null,
        account_holder: result.accountHolder
          ? {
              name: result.accountHolder.name || null,
              match_status: result.accountHolder.matchStatus,
            }
          : null,
        alternatives: result.alternatives.map((alternative) => ({
          account_number: alternative.accountNumber,
          bank_code: alternative.bankCode,
          bank_name: alternative.bankName,
          confidence: alternative.confidence,
          nuban_valid: alternative.nubanValid,
        })),
      },
      meta: {
        confidence: result.confidence,
        processing_time_ms: result.processingTimeMs,
        source: 'ocr',
      },
    });
  } catch (error) {
    handleError(error, res);
  }
});

router.post('/recognize/base64', async (req: Request, res: Response) => {
  try {
    const { image_base64, image_format, options = {} } = req.body;

    if (!image_base64) {
      res.status(400).json({
        status: 'error',
        error: {
          code: 'INVALID_REQUEST',
          message: 'Missing image_base64 field in request body.',
        },
      });
      return;
    }

    const imageBuffer = Buffer.from(image_base64, 'base64');

    if (imageBuffer.length > config.maxImageSize) {
      throw new ImageTooLargeError(imageBuffer.length);
    }

    const metadata = await sharp(imageBuffer).metadata();
    if (
      (metadata.width && metadata.width < config.minImageWidth) ||
      (metadata.height && metadata.height < config.minImageHeight)
    ) {
      throw new ImageTooSmallError(metadata.width || 0, metadata.height || 0);
    }

    const recognizeOptions = options as RecognizeOptions;
    const result = await recognize(imageBuffer, {
      ...recognizeOptions,
      returnAlternatives: true,
      maxAlternatives: recognizeOptions.maxAlternatives ?? 20,
    });

    res.json({
      status: 'success',
      data: {
        account_number: result.accountNumber,
        bank_code: result.bank?.code || null,
        bank_name: result.bank?.name || null,
        account_name: result.accountHolder?.name || null,
        account_holder: result.accountHolder
          ? {
              name: result.accountHolder.name || null,
              match_status: result.accountHolder.matchStatus,
            }
          : null,
        alternatives: result.alternatives.map((alternative) => ({
          account_number: alternative.accountNumber,
          bank_code: alternative.bankCode,
          bank_name: alternative.bankName,
          confidence: alternative.confidence,
          nuban_valid: alternative.nubanValid,
        })),
      },
      meta: {
        confidence: result.confidence,
        processing_time_ms: result.processingTimeMs,
        source: 'ocr',
      },
    });
  } catch (error) {
    handleError(error, res);
  }
});

function handleError(error: unknown, res: Response): void {
  
  if ((error as any)?.code === 'LIMIT_FILE_SIZE') {
    res.status(400).json({
      status: 'error',
      error: {
        code: 'IMAGE_TOO_LARGE',
        message: `Image exceeds the maximum file size of ${config.maxImageSize / (1024 * 1024)} MB.`,
      },
    });
    return;
  }

  if (error instanceof SnapAccountError) {
    res.status(error.httpStatus).json({
      status: 'error',
      error: {
        code: error.code,
        message: error.message,
        details: error.details,
      },
    });
    return;
  }

  console.error('[Recognize] Unexpected error:', error);
  res.status(500).json({
    status: 'error',
    error: {
      code: 'INTERNAL_ERROR',
      message: 'An unexpected error occurred during recognition.',
    },
  });
}

export default router;

File: snapaccount-py/routes/recognize.py


import base64
import io

from fastapi import APIRouter, UploadFile, File, HTTPException
from pydantic import BaseModel
from typing import Optional
from PIL import Image

from config import API_VERSION, MAX_IMAGE_SIZE_BYTES, MIN_IMAGE_WIDTH, MIN_IMAGE_HEIGHT
from services.pipeline import run

router = APIRouter()

class Base64Request(BaseModel):
    image: str  
    hint: Optional[str] = None  

def _format_result(result) -> dict:
    """Convert a RecognitionResult to the Node.js compatible API response format."""
    if result.error_code:
        return {
            "status": "error",
            "error": {
                "code": result.error_code,
                "message": result.error_message,
                "details": {
                    "suggestion": "Retake the photo with better lighting or crop the image.",
                    "image_quality": result.image_quality,
                    "processing_time_ms": result.processing_time_ms,
                    "fallback": "manual_entry"
                }
            }
        }

    return {
        "status": "success",
        "data": {
            "account_number": result.account_number,
            "bank_code": result.bank_code,
            "bank_name": result.bank_name,
            "account_name": result.account_name,
        },
        "meta": {
            "confidence": result.confidence,
            "processing_time_ms": result.processing_time_ms,
            "source": "ocr"
        }
    }

def _validate_image_bytes(image_bytes: bytes) -> None:
    """Validate image size and dimensions."""
    if len(image_bytes) > MAX_IMAGE_SIZE_BYTES:
        raise HTTPException(
            status_code=400,
            detail={
                "status": "error",
                "error_code": "IMAGE_TOO_LARGE",
                "message": f"Image exceeds maximum size of {MAX_IMAGE_SIZE_BYTES // (1024 * 1024)} MB.",
            },
        )

    try:
        img = Image.open(io.BytesIO(image_bytes))
        width, height = img.size
        if width < MIN_IMAGE_WIDTH or height < MIN_IMAGE_HEIGHT:
            raise HTTPException(
                status_code=400,
                detail={
                    "status": "error",
                    "error_code": "IMAGE_TOO_SMALL",
                    "message": f"Image resolution ({width}×{height}) is below minimum ({MIN_IMAGE_WIDTH}×{MIN_IMAGE_HEIGHT}).",
                },
            )
    except HTTPException:
        raise
    except Exception:
        raise HTTPException(
            status_code=400,
            detail={
                "status": "error",
                "error_code": "IMAGE_FORMAT_UNSUPPORTED",
                "message": "Could not read image. Accepted formats: JPEG, PNG, WebP.",
            },
        )

@router.post("/recognize")
async def recognize_multipart(image: UploadFile = File(...)):
    """
    POST /v1/recognize — multipart image upload.

    Send a photo as multipart/form-data with field name "image".
    Returns the recognised account number, bank, and holder name.
    """
    image_bytes = await image.read()
    _validate_image_bytes(image_bytes)

    result = await run(image_bytes)
    return _format_result(result)

@router.post("/recognize/base64")
async def recognize_base64(body: Base64Request):
    """
    POST /v1/recognize/base64 — JSON with base64-encoded image.

    Send {"image": "<base64>", "hint": "GTBank"} in JSON body.
    Useful for server-to-server integrations.
    """
    try:
        image_bytes = base64.b64decode(body.image)
    except Exception:
        raise HTTPException(
            status_code=400,
            detail={
                "status": "error",
                "error_code": "INVALID_REQUEST",
                "message": "Invalid base64 encoding for image field.",
            },
        )

    _validate_image_bytes(image_bytes)

    result = await run(image_bytes)
    return _format_result(result)
